istioctl/pkg/genbinding/genbinding.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package genbinding
+
+import (
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+type hostPort struct {
+	host string
+	port int
+}
+
+const defaultMultiMeshPort = "15443"
+
+var validHostnameRegex = regexp.MustCompile(`^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])$`)
+
+// CreateBinding converts desired multicluster state into Istio state
+func CreateBinding(service string, clusters []string, labels map[string]string, egressGateway string, namespace string) ([]model.Config, error) { // nolint: lll
+
+	remoteService, err := parseNet(service, "")
+	if err != nil {
+		return nil, err
+	}
+
+	var remoteClusters []hostPort
+	for _, cluster := range clusters {
+		remoteCluster, err := parseNet(cluster, defaultMultiMeshPort)
+		if err != nil {
+			return nil, err
+		}
+		remoteClusters = append(remoteClusters, *remoteCluster)
+	}
+
+	istioConfig, err := serviceToServiceEntrySniCluster(*remoteService, remoteClusters, labels, egressGateway, namespace)
+	return istioConfig, err
+}
+
+// serviceToServiceEntry() creates a ServiceEntry pointing to istio-egressgateway
+func serviceToServiceEntrySniCluster(remoteService hostPort, remoteClusters []hostPort, labels map[string]string, egressGateway string, namespace string) ([]model.Config, error) { // nolint: lll
+	protocol := "http"
+	var egresslabels map[string]string
+	// It is strongly recommended addresses for different hosts don't clash;
```

Comment:
Isn't it more than "strongly recommended", i.e., "required"? If this ever generates a clashing IP, it will not work. Maybe the user needs to pass in the IP address to use, or otherwise some context with which the command can allocate an IP that is guaranteed not to clash with any other one.

Reply:
There is no good way to do this from _istioctl_.  Every solution requires lots of manual fiddling or imitating transactional behavior.

A better approach would be to have a mutating webhook look for "multicluster" ServiceEntries and generate a unique 127.255.x.y IP address for them.  Perhaps we should temporarily table this PR and pivot to creating that functionality first?

Reply:
Or just ask users to pass in the IP. :). to keep it simple for now.

Reply:
Yes, now you can specify an IP address to be used for the vip. If such an address is not specified, the service entry template will be created without an address. In such a case, a simple solution such as the one proposed here: https://github.com/istio-ecosystem/istio-coredns-plugin/pull/3 can assign the address as run time.